### PR TITLE
show generic parameter sig help after conditional access

### LINKED ISSUE #1598: System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.MemberBindingExpressionSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.MemberAccessExpressionSyntax'.
I had an array of objects and I was in the middle of typing ```?.OfType<DynamicAttribute>()```.  When ```OfType``` failed to bind (before typing "<"?) I manually added ```using System.Linq;```.  Then I returned to typing the type argument.  I think I had probably ctrl-. after "D".

```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.MemberBindingExpressionSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.MemberAccessExpressionSyntax'.
   at Microsoft.CodeAnalysis.CSharp.Extensions.SimpleNameSyntaxExtensions.GetLeftSideOfDot(SimpleNameSyntax name)
   at Microsoft.CodeAnalysis.Editor.CSharp.SignatureHelp.GenericNameSignatureHelpProvider.<GetItemsWorkerAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.AbstractSignatureHelpProvider.<GetItemsAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller.Session.<ComputeItemsAsync>d__10.MoveNext()
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.ReportUnlessCanceled(System.Exception)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session+<ComputeItemsAsync>d__10.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session+<ComputeItemsAsync>d__10.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Start[[Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session+<ComputeItemsAsync>d__10, Microsoft.CodeAnalysis.EditorFeatures, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ComputeItemsAsync>d__10 ByRef)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session.ComputeItemsAsync(System.Collections.Generic.IList`1<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider>, Microsoft.VisualStudio.Text.SnapshotPoint, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo, Microsoft.CodeAnalysis.Document, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session+<ComputeModelInBackgroundAsync>d__5.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Start[[Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session+<ComputeModelInBackgroundAsync>d__5, Microsoft.CodeAnalysis.EditorFeatures, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ComputeModelInBackgroundAsync>d__5 ByRef)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session.ComputeModelInBackgroundAsync(Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Model, System.Collections.Generic.IList`1<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider>, System.Collections.Generic.IList`1<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider>, Microsoft.VisualStudio.Text.SnapshotPoint, Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.DisconnectedBufferGraph, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller+Session+<>c__DisplayClass4_0.<ComputeModel>b__0(Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Model, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.ModelComputation`1+<>c__DisplayClass17_0[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<ChainTaskAndNotifyControllerWhenFinished>b__0(System.Threading.Tasks.Task`1<System.__Canon>)
   at Roslyn.Utilities.TaskExtensions+<>c__DisplayClass14_0`2[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<SafeContinueWithFromAsync>b__0(System.Threading.Tasks.Task)
   at System.Threading.Tasks.ContinuationResultTaskFromTask`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean)
   at System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

### PR DESCRIPTION
**Pull request by brettfo**
Fixes #1598

This is a crashing bug and is a candidate for preview.

Note that bug #1628 was found during testing, but since it's non-crashing that bugfix will not be taken in the preview branch.

### CODE REVIEW

**Comment by brettfo** [comment]
Tagging @Pilchie @dpoeschl @jasonmalinowski @rchande @balajikris @basoundr as reviewers for a preview bugfix.

**Comment by Pilchie** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/Extensions/SimpleNameSyntaxExtensions.cs [hunk 1]
```diff
@@ -15,7 +15,15 @@ public static ExpressionSyntax GetLeftSideOfDot(this SimpleNameSyntax name)
...
-                return ((MemberAccessExpressionSyntax)name.Parent).Expression;
+                var conditionalAccess = name.Parent.Parent as ConditionalAccessExpressionSyntax;
+                if (conditionalAccess != null)
+                {
+                    return conditionalAccess.Expression;
+                }
+                else
+                {
+                    return ((MemberAccessExpressionSyntax)name.Parent).Expression;
+                }
```
Comment (rchande) [inline src/Workspaces/CSharp/Portable/Extensions/SimpleNameSyntaxExtensions.cs]: Use SyntaxNodeExtensions.GetParentConditionalAccessExpression. It always finds the right node, even with more complicated conditional access.

File: src/Workspaces/CSharp/Portable/Extensions/SimpleNameSyntaxExtensions.cs [hunk 2]
```diff
@@ -15,7 +15,15 @@ public static ExpressionSyntax GetLeftSideOfDot(this SimpleNameSyntax name)
             Contract.Requires(name.IsMemberAccessExpressionName() || name.IsRightSideOfQualifiedName() || name.IsParentKind(SyntaxKind.NameMemberCref));
             if (name.IsMemberAccessExpressionName())
             {
-                return ((MemberAccessExpressionSyntax)name.Parent).Expression;
+                var conditionalAccess = name.Parent.Parent as ConditionalAccessExpressionSyntax;
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/SimpleNameSyntaxExtensions.cs]: Is this only an issue if the `ConditionalAccessExpressionSyntax` is the grandparent?  What if it's further up a chain of dotting?

File: src/EditorFeatures/Test2/IntelliSense/CSharpSignatureHelpCommandHandlerTests.vb [hunk 3]
```diff
@@ -356,5 +358,28 @@ class C
...
+
+class C
+{
+    void M(object[] args)
+    {
+        var x = args?.OfType$$
+    }
+}
+]]></Document>)
+
```
Comment (rchande) [inline src/EditorFeatures/Test2/IntelliSense/CSharpSignatureHelpCommandHandlerTests.vb]: Tests for more complicated cases? :smile:

**Comment by Pilchie** [comment]
Also, make sure to run the tests not covered by Jenkins, in case that extension is used by other features.

**Comment by balajikris** [comment]
I know VB parses this slightly differently, but have you checked that it doesn't crash in VB too?

File: src/EditorFeatures/Test2/IntelliSense/CSharpSignatureHelpCommandHandlerTests.vb (hunk 3 above)
Comment (balajikris) [inline src/EditorFeatures/Test2/IntelliSense/CSharpSignatureHelpCommandHandlerTests.vb]: like nested conditional access expressions..

**Comment by rchande** [comment]
In case you need it, VB has the (arbitrarily differently named) GetCorrespondingConditionalAccessExpression helper.

**Comment by brettfo** [comment]
I've switched to use `SyntaxNodeExtensions.GetParentConditionalAccessExpression` as suggested, added more complicated C# tests, and I also added VB tests to ensure there's no regression (no fix for VB was needed.)

I've also run all open and closed tests on my box.

**Comment by rchande** [comment]
:+1:

**Comment by Pilchie** [comment]
:+1:

**Pull request merged**